apps/src/studio/studio.js

```diff
@@ -2478,13 +2523,60 @@ Studio.createLevelItems = function (svg) {
   }
 };
 
-Studio.drawMapTiles = function () {
+Studio.isWallTile = function(row, col) {
+  if (row < 0 || row >= Studio.ROWS || col < 0 || col >= Studio.COLS) {
+    return false;
+  }
+
+  return (Studio.map[row][col] & SquareType.WALL || (Studio.walls !== null && skin[Studio.walls] && skin[Studio.walls][row][col]));
+};
+
+Studio.getWallTile = function(row, col) {
+  if (Studio.walls !== null) {
+    return skin[Studio.walls][row][col];
+  } else if (Studio.map[row][col] & SquareType.WALL) {
+    return 1;
+  }
+};
+
+Studio.drawMapTiles = function (svg) {
+  var row, col;
+
+  var tilesDrawn = [];
+  for (row = 0; row < Studio.ROWS; row++) {
+    tilesDrawn[row] = [];
+    for (col = 0; col < Studio.COLS; col++) {
+      tilesDrawn[row][col] = false;
+    }
+  }
+
   var spriteLayer = document.getElementById('backgroundLayer');
-  for (var row = 0; row < Studio.ROWS; row++) {
-    for (var col = 0; col < Studio.COLS; col++) {
-      var wallVal = Studio.getWallValue(row, col);
-      if (wallVal) {
-        Studio.drawWallTile(spriteLayer, wallVal, row, col);
+
+  for (row = 0; row < Studio.ROWS; row++) {
+    for (col = 0; col < Studio.COLS; col++) {
+      if (Studio.isWallTile(row, col)) {
+        // Skip if we've already drawn a large tile that covers this square.
+        if (tilesDrawn[row][col]) {
+          continue;
+        }
+
+        var largeTile = false;
+
+        // If we can draw a large tile here, then do so, and make sure
+        // we won't draw any more tiles where it extends over.
+        if (Studio.isWallTile(row, col+1) &&
+            Studio.isWallTile(row+1, col) &&
+            Studio.isWallTile(row+1, col+1)) {
+          largeTile = true;
```

Comment:
I can't figure out how the map author can decide whether or not a large tile is rendered. It looks like anytime we see a contiguous block of 2x2, we'll go ahead and create a large tile. That could make sense when we're placing random wall tiles. But for the explicit walls, wouldn't it be better to give control to the map author?